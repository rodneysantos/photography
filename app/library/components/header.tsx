import { clsx } from "clsx";
import { useState } from "react";
import { ChevronDownIcon } from "./chevron-down-icon";
import { Item, Menu } from "./menu";

interface HeaderProps {
	// The function that is called when the active items change.
	onActiveItemsChange: (activeItems: Item[]) => void;
}

export const Header = ({ onActiveItemsChange }: HeaderProps) => {
	const [isMenuOpen, setIsMenuOpen] = useState<boolean>(false);

	// The useCallback hook is used to memoize the function handleMenuVisibility
	// so that it is only re-created when the value of isMenuOpen changes.
	const handleMenuVisibility = () => {
		setIsMenuOpen((prevState) => !prevState);
	};

	return (
		<div className="flex flex-col mt-4 text-center">
			<div className="z-10 mb-4 bg-white">
				<span className="text-4xl tracking-widest font-big-shoulder">
					RODNEY SANTOS
				</span>

				<br />

				<span className="ml-[0.55em] font-dosis text-base tracking-[0.83em]">
					PHOTOGRAPHY
				</span>
			</div>

			<div
				data-testid="menuContainer"
				className={`flex flex-col items-center transition-all will-change-transform ${clsx(
					{
						"-translate-y-14": !isMenuOpen,
					},
				)}`}
			>
				<Menu
					defaultActiveItems={["black and white", "low-key"]}
					onActiveItemsChange={(activeItems) =>
						onActiveItemsChange(activeItems)
					}
					className={`transition-opacity will-change-auto ${clsx({
						"opacity-0": !isMenuOpen,
					})}`}
				/>

				<ChevronDownIcon
					className={`w-5 mt-4 transition-opacity cursor-pointer will-change-auto ${clsx(
						{
							"opacity-20": !isMenuOpen,
						},
					)}`}
					onClick={handleMenuVisibility}
				/>
			</div>
		</div>
	);
};
