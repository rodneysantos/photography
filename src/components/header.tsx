import { createSignal } from "solid-js";
import { ChevronDownIcon } from "./chevron-down-icon";
import { MenuKey, Menu } from "./menu";

interface HeaderProps {
	// The function that is called when the active items change.
	onActiveItemsChange: (activeItems: MenuKey[]) => void;
}

export const Header = ({ onActiveItemsChange }: HeaderProps) => {
	const [isMenuOpen, setIsMenuOpen] = createSignal(false);

	// The useCallback hook is used to memoize the function handleMenuVisibility
	// so that it is only re-created when the value of isMenuOpen changes.
	const handleMenuVisibility = () => {
		setIsMenuOpen((prevState) => !prevState);
	};

	return (
		<div class="flex flex-col text-center mt-14 5xl:mt-16 5xl:mb-4">
			<div class="z-10 mb-4 bg-white">
				<div class="tracking-widest text-3xl font-big-shoulder">
					RODNEY SANTOS
				</div>

				<div class="text-xl tracking-xs translate-x-1/10 font-dosis">
					PHOTOGRAPHY
				</div>
			</div>

			<div
				data-testid="menuContainer"
				class="flex flex-col items-center transition-all will-change-transform"
				classList={{ "-translate-y-14": !isMenuOpen() }}
			>
				<Menu
					defaultActiveItems={["bnw", "low-key"]}
					onActiveItemsChange={onActiveItemsChange}
					show={isMenuOpen()}
				/>

				<ChevronDownIcon onClick={handleMenuVisibility} show={!isMenuOpen()} />
			</div>
		</div>
	);
};
