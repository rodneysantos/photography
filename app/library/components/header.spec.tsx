import { act, fireEvent, render } from "@testing-library/react";
import renderer from "react-test-renderer";
import { beforeEach, describe, expect, it, vi } from "vitest";
import { Header } from "./header";

describe("header", () => {
	const onActiveItemsChange = vi.fn();

	beforeEach(() => {
		onActiveItemsChange.mockReset();
	});

	it("renders correctly", () => {
		// arrange
		const tree = renderer
			.create(<Header onActiveItemsChange={onActiveItemsChange} />)
			.toJSON();

		// assert
		expect(tree).toMatchSnapshot();
	});

	it("animates the menu container on click", () => {
		// arrange
		const { getByTestId } = render(
			<Header onActiveItemsChange={onActiveItemsChange} />,
		);

		// act
		const container = getByTestId("menuContainer");
		const dropdown = getByTestId("chevronDownIcon");
		act(() => {
			fireEvent.click(dropdown);

			// assert
			expect(container.className).toContain("-translate-y-14");
			expect(dropdown.className).toContain("opacity-20");
		});
	});

	it("emits the active items on change", () => {
		// arrange
		const { getByTestId } = render(
			<Header onActiveItemsChange={onActiveItemsChange} />,
		);

		// act
		const blackAndWhiteItem = getByTestId("black and white");
		act(() => {
			blackAndWhiteItem.click();
		});

		// assert
		expect(onActiveItemsChange).toHaveBeenCalledTimes(1);
	});
});
