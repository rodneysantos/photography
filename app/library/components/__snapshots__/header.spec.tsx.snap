// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`header > renders correctly 1`] = `
<div
  className="flex flex-col mt-4 text-center"
>
  <div
    className="z-10 mb-4 bg-white"
  >
    <span
      className="text-4xl tracking-widest font-big-shoulder"
    >
      RODNEY SANTOS
    </span>
    <br />
    <span
      className="ml-[0.55em] font-dosis text-base tracking-[0.83em]"
    >
      PHOTOGRAPHY
    </span>
  </div>
  <div
    className="flex flex-col items-center transition-all will-change-transform -translate-y-14"
    data-testid="menuContainer"
  >
    <div
      className="transition-opacity will-change-auto opacity-0"
      data-testid="menu"
    >
      <div
        className="flex flex-row justify-center"
      >
        <div
          className="px-2 py-1 mx-2 text-center border-2 cursor-pointer border-neutral-900 bg-neutral-900 text-neutral-100"
          data-testid="black and white"
          onClick={[Function]}
        >
          black and white
        </div>
        <div
          className="px-2 py-1 mx-2 text-center border-2 cursor-pointer border-neutral-900 "
          data-testid="color"
          onClick={[Function]}
        >
          color
        </div>
        <div
          className="px-2 py-1 mx-2 text-center border-2 cursor-pointer border-neutral-900 bg-neutral-900 text-neutral-100"
          data-testid="low-key"
          onClick={[Function]}
        >
          low-key
        </div>
      </div>
    </div>
    <svg
      className="w-5 mt-4 transition-opacity cursor-pointer will-change-auto opacity-20"
      data-testid="chevronDownIcon"
      onClick={[Function]}
      onKeyDown={[Function]}
      viewBox="0 0 512 512"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M256 0a256 256 0 1 0 0 512A256 256 0 1 0 256 0zM135 241c-9.4-9.4-9.4-24.6 0-33.9s24.6-9.4 33.9 0l87 87 87-87c9.4-9.4 24.6-9.4 33.9 0s9.4 24.6 0 33.9L273 345c-9.4 9.4-24.6 9.4-33.9 0L135 241z"
      />
    </svg>
  </div>
</div>
`;
